### PR TITLE
Issue/5416 refunding fees

### LINKED ISSUE #5416: Add support for refunding fees
Ref p91TBi-6Px-p2 p91TBi-6Ou-p2

Support refunding fees. Possibly bundle all the fees together. Add a toggle just like the one used for refunding shipping fees.

- [ ] Show the fees line in the Issues Refund screen
- [ ] Only allow if the fees have not been previously refunded
- [ ] Make sure Order Detail fees are showing bundled fees too

iOS ticket - https://github.com/woocommerce/woocommerce-ios/issues/5624

### PR DESCRIPTION
**Pull request by malinajirka**
Closes: #5416 

Merge Instructions:
1. [Review PR in FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2226)
2. Review this PR
3. Merge FluxC PR
4. Update FluxC hash + remove "Do not merge" label
5. Merge this PR

### Description

Goal of this PR is to add support for refunding fees from the app.

I'm sorry for the size of this PR. I'm usually against PRs of this size, however, 90+% of the code is a simply copy-paste of the logic for shipping refunds. **I recommend reviewing this PR by commits to follow my train of thoughts.**

Commits 00c27d87782d018730130eb329f5362191d4b1c6 and fbb51eb9470390a3e8a9ef029ce3c4f79074faf2 are by far the largest, but they contain only a copy-paste of shipping refunds logic + renaming. They should still be reviewed, since I might have made a mistake, but it's not necessary to review the logic.

Notes:
1. The code contains even a RecyclerView (including the adapter, layouts, view logic) that lists all the FeeLines of the order. However, the visibility of this recycler view is hard coded to GONE so the code is technically never used. I wasn't sure if I should include a dead code. However, I opted for creating a copy of the behavior of Shipping refunds. If we decide to support refunding fees by row, it'll be just a matter of making the RecyclerView and Switcher on list items visible. I plan to create a project proposal which will consolidate all this code - either remove it or enable the support.
2. I've noticed a few issues with Refunds when I was working on this PR - I'll elaborate on these on the project proposal.
- "Issue Refund" button is not visible when only shipping has not been refunded yet 
- "You can refund taxes in your store admin" label is visible. However, the app allows the user to refund taxes - all shipping/fee/product taxes. It causes some unexpected side-effects in wpadmin.
- Product refund section is visible even when the order doesn't have any products - eg. it contains only fees (such as Simple Payments Order)
3. I added just a very basic unit tests - the VM is missing most of the unit tests and I plan to cover that in the project proposal too

### Testing instructions

1. Open detail of an order with a fee item available for refund
2. Tap on "Issue Refunds" button
3. Verify that the Refund Fees Section is displayed as a single row containing a disabled toggle
4. Activate the Toggle and verify that the full Refund Fees section shows up, displaying currently the subtotal, tax and total fields, also verify that the Next button becomes clickable too.
5. Hit the Next button and verify if the Issue Confirmation view show up correctly as expected, hit the Refund button and check if the Refund happened as expected

- Repeat the same steps as above on an order with multiple fee items

### Images/gif

| 1 | 2 |
| --- | --- |
| Screenshot_1640265239 | Screenshot_1640265240 |

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/res/layout/refund_fee_list_item.xml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (malinajirka) [inline WooCommerce/src/main/res/layout/refund_fee_list_item.xml:1]: This file is not used as mentioned in the desription.

File: WooCommerce/src/main/res/layout/refund_by_items_fees.xml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        android:layout_height="wrap_content"
+        android:orientation="vertical">
+
+        <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/issueRefund_feeLines"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:nestedScrollingEnabled="false"
+            tools:itemCount="2"
+            android:visibility="gone"
```
Comment (malinajirka) [inline WooCommerce/src/main/res/layout/refund_by_items_fees.xml:21]: This RV is not used as mentioned in the desription.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundFeeListAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import com.google.android.material.switchmaterial.SwitchMaterial
+import com.woocommerce.android.R
+import com.woocommerce.android.extensions.hide
+import com.woocommerce.android.model.Order
+import kotlinx.android.parcel.Parcelize
+import org.wordpress.android.fluxc.model.refunds.WCRefundModel.WCRefundItem
+import java.math.BigDecimal
+
+class RefundFeeListAdapter(
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundFeeListAdapter.kt:17]: This file is not used as mentioned in the desription.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Refund.kt [hunk 4]
```diff
@@ -99,6 +108,15 @@ fun WCRefundShippingLine.toAppModel(): Refund.ShippingLine {
     )
 }
 
+fun WCRefundFeeLine.toAppModel(): Refund.FeeLine {
+    return Refund.FeeLine(
+        id = metaData?.get(0)?.value?.toString()?.toLongOrNull() ?: -1,
+        name = name,
+        totalTax = -totalTax.roundError(), // WCRefundFeeLine.totalTax is NEGATIVE
+        total = (total).roundError(), // WCRefundFeeLine.total is NEGATIVE
```
Comment (malinajirka) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/Refund.kt:116]: ⚠️ It seems this line doesn't have `-` at the beginning even though the comment says it's negative. I copied the code from shipping but I'm wondering if it's a bug or an intentional thing.

Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/Refund.kt:116]: Thanks for mentioning this. I've completely forgotten why they're this way for shipping line, and I couldn't find anything in my notes either.

My general feel though is that the whole refund shipping/fee/extra things could use a more thorough look and polish. The shipping refund code was one of my earliest introduction to the Woo mobile codebase so there's definitely some cruft in there (e.g: about how multiple shipping lines refund isn't yet supported). I'd be looking forward to your project proposal and hopefully I can contribute as well.

**Review by hafizrahman — APPROVED** [review]
Code works and make sense to me, especially as you mentioned that quite a bit of them were similar to the shipping lines refund.

As mentioned on the above comment, I'm looking forward to your project proposal to make this area better, and for now functionality-wise this PR should suffice. The similarity between this and shipping refund hopefully will make them both easier to refactor, if needed!

**Pull request merged**